File: dotnet/src/RocketMQ.Client/Common/SubscriptionData.cs
// 订阅数据（对应 org.apache.rocketmq.remoting.protocol.heartbeat.SubscriptionData
// 与 org.apache.rocketmq.common.filter.ExpressionType / FilterAPI）。
//
// 用途：Consumer 注册到 broker 的心跳里携带 SubscriptionData，broker 依据
// tagsSet / codeSet 做 TAG 过滤；subString 为 "||" 分隔的 tag 表达式。
//
// JSON 字段（与 fastjson2 序列化结果一致，键按字母序）：
//   classFilterMode | codeSet | expressionType | subString | subVersion | tagsSet | topic
// 注意：Java 的 filterClassSource 标注了 @JSONField(serialize = false)，**不参与**序列化。
using System;
using System.Collections.Generic;
using System.Globalization;
using RocketMQ.Remoting.Protocol;

namespace RocketMQ.Common;

// org.apache.rocketmq.common.filter.ExpressionType
public static class ExpressionType
{
    public const string TAG = "TAG";
    public const string Sql92 = "SQL92";
    public const string ClassFilter = "CLASS_FILTER";
}

public sealed class SubscriptionData : IComparable<SubscriptionData>
{
    public bool ClassFilterMode { get; set; }
    public string Topic { get; set; } = string.Empty;
    public string SubString { get; set; } = string.Empty;
    public SortedSet<string> TagsSet { get; set; } = new SortedSet<string>(StringComparer.Ordinal);
    public SortedSet<long> CodeSet { get; set; } = new SortedSet<long>();
    public long SubVersion { get; set; }
    public string ExpressionType { get; set; } = RocketMQ.Common.ExpressionType.TAG;
    // 仅本地使用，不参与 JSON（对齐 Java @JSONField(serialize = false)）
    public string FilterClassSource { get; set; } = string.Empty;

    public SubscriptionData()
    {
        // 默认取当前毫秒时间戳（C++ 默认构造函数行为一致）
        SubVersion = UtilAll.CurrentTimeMillis();
    }

    public SubscriptionData(string topic, string subString)
    {
        Topic = topic;
        SubString = subString;
        SubVersion = UtilAll.CurrentTimeMillis();
    }

    // ---- JSON ----
    // 键按字母序，与 fastjson2 输出一致
    public JsonValue ToJson()
    {
        JsonValue o = JsonValue.MakeObject();
        o.Set("classFilterMode", JsonValue.MakeBool(ClassFilterMode));
        o.Set("codeSet", SetToJson(CodeSet));
        o.Set("expressionType", JsonValue.MakeString(ExpressionType));
        o.Set("subString", JsonValue.MakeString(SubString));
        o.Set("subVersion", JsonValue.MakeInt(SubVersion));
        o.Set("tagsSet", SetToJson(TagsSet));
        o.Set("topic", JsonValue.MakeString(Topic));
        return o;
    }

    public static SubscriptionData FromJson(JsonValue v)
    {
        SubscriptionData sd = new SubscriptionData();
        if (!v.IsObject)
        {
            return sd;
        }

        if (v.TryGetBool("classFilterMode", out bool b))
        {
            sd.ClassFilterMode = b;
        }

        if (v.TryGetString("expressionType", out string s1))
        {
            sd.ExpressionType = s1;
        }

        if (v.TryGetString("subString", out string s2))
        {
            sd.SubString = s2;
        }

        if (v.TryGetString("topic", out string s3))
        {
            sd.Topic = s3;
        }

        if (v.TryGetInt("subVersion", out long n))
        {
            sd.SubVersion = n;
        }

        JsonValue? tags = v.Find("tagsSet");
        if (tags is not null && tags.IsArray)
        {
            for (int i = 0; i < tags.Size(); i++)
            {
                sd.TagsSet.Add(tags.At(i).StringValue());
            }
        }

        JsonValue? codes = v.Find("codeSet");
        if (codes is not null && codes.IsArray)
        {
            for (int i = 0; i < codes.Size(); i++)
            {
                sd.CodeSet.Add(codes.At(i).IntValue());
            }
        }

        return sd;
    }

    // Java equals 语义：比较 classFilterMode/codeSet/subString/subVersion/tagsSet/
    // topic/expressionType（**含** subVersion；不含 filterClassSource）。
    public override bool Equals(object? obj)
    {
        if (obj is not SubscriptionData other)
        {
            return false;
        }

        return ClassFilterMode == other.ClassFilterMode
            && CodeSet.SetEquals(other.CodeSet)
            && SubString == other.SubString
            && SubVersion == other.SubVersion
            && TagsSet.SetEquals(other.TagsSet)
            && Topic == other.Topic
            && ExpressionType == other.ExpressionType;
    }

    // Java compareTo：按 "topic@subString" 字符串序
    public int CompareTo(SubscriptionData? other)
    {
        if (other is null)
        {
            return 1;
        }

        string a = Topic + "@" + SubString;
        string b = other.Topic + "@" + other.SubString;
        if (a == b)
        {
            return 0;
        }

        return a.CompareTo(b) < 0 ? -1 : 1;
    }

    public override int GetHashCode()
    {
        // Java int/long 溢出语义：结果按 32 位有符号回绕
        long result = 1;
        result = 31 * result + (ClassFilterMode ? 1231 : 1237);
        result = 31 * result + SetHashCode(CodeSet);
        result = 31 * result + JavaHash.JavaStringHash(SubString);
        result = 31 * result + SetHashCode(TagsSet);
        result = 31 * result + JavaHash.JavaStringHash(Topic);
        result = 31 * result + JavaHash.JavaStringHash(ExpressionType);
        result &= 0xFFFFFFFFL;
        return unchecked((int)result);
    }

    public static bool operator ==(SubscriptionData? left, SubscriptionData? right)
    {
        if (left is null)
        {
            return right is null;
        }

        return left.Equals(right);
    }

    public static bool operator !=(SubscriptionData? left, SubscriptionData? right) => !(left == right);

    public override string ToString()
    {
        return "SubscriptionData [classFilterMode=" + (ClassFilterMode ? "true" : "false")
            + ", topic=" + Topic + ", subString=" + SubString
            + ", tagsSet=" + SetToJson(TagsSet).Dump()
            + ", codeSet=" + SetToJson(CodeSet).Dump()
            + ", subVersion=" + SubVersion.ToString(CultureInfo.InvariantCulture)
            + ", expressionType=" + ExpressionType + "]";
    }

    // ---- 内部辅助 ----
    // Java Set.hashCode() = 各元素 hashCode 之和（32 位有符号回绕）
    private static int SetHashCode(SortedSet<string> s)
    {
        long h = 0;
        foreach (string e in s)
        {
            h += JavaHash.JavaStringHash(e);
            h &= 0xFFFFFFFFL;
        }

        return unchecked((int)h);
    }

    private static int SetHashCode(SortedSet<long> s)
    {
        long h = 0;
        foreach (long e in s)
        {
            h += e;
            h &= 0xFFFFFFFFL;
        }

        return unchecked((int)h);
    }

    // Java set 序列化为 JSON 数组
    private static JsonValue SetToJson(SortedSet<string> s)
    {
        JsonValue arr = JsonValue.MakeArray();
        foreach (string e in s)
        {
            arr.PushArray(JsonValue.MakeString(e));
        }

        return arr;
    }

    private static JsonValue SetToJson(SortedSet<long> s)
    {
        JsonValue arr = JsonValue.MakeArray();
        foreach (long e in s)
        {
            arr.PushArray(JsonValue.MakeInt(e));
        }

        return arr;
    }
}

// org.apache.rocketmq.common.filter.FilterAPI
public static class FilterAPI
{
    public const string SubAll = "*";

    // 对齐 Java FilterAPI.buildSubscriptionData（Java 探针实测向量）：
    //   null / "" / "*"   → subString 归一为 "*"，**tagsSet 与 codeSet 都保持空**
    //   "TagA"            → tagsSet={TagA}, codeSet={2598919}
    //   "TagA||TagB"      → tagsSet={TagA,TagB}, codeSet={2598919,2598920}
    //   " TagA || TagB "  → subString 原样保留空格，标签各自 trim
    //   "   "（纯空白）    → tagsSet 空、subString 原样保留（IsEmpty 只认 null/""）
    //   "|||"             → tagsSet={|}, codeSet={124}（Java-split 只丢**末尾**空串）
    //   "||" / "||||"     → 抛 "subString split error"（Java-split 结果数组长度为 0）
    //
    // ⚠ 两处曾有的偏差（都会污染心跳、并让客户端二次 tag 过滤失效）：
    //   ① 给 "*" 塞 tagsSet={"*"} —— Java 里 tagsSet 非空才是"客户端二次 tag 过滤"的开关
    //      （PullAPIWrapper.processPullResult 的 `!tagsSet.isEmpty()`），塞了 "*" 会让订阅
    //      全量时把所有正常 tag 的消息客户端自己过滤掉；
    //   ② 从不填 codeSet —— 它是 broker 侧按 tag 哈希过滤的依据
    //      （ExpressionMessageFilter.isMatchedByConsumeQueue 走 codeSet.contains）。
    // 另注：判空必须用"空串"而不是"全空白"—— Java StringUtils.isEmpty 只认 null/""，
    // 纯空白（如 "   "）会走进 split 分支，结果 tagsSet 空但 subString 原样保留。
    public static SubscriptionData BuildSubscriptionData(string topic, string subString)
    {
        SubscriptionData sub = new SubscriptionData(topic, subString);
        if (string.IsNullOrEmpty(subString) || subString == SubAll)
        {
            sub.SubString = SubAll;
            return sub;
        }

        // Java String.split("\\|\\|")：先全切，再丢弃**末尾**空串
        List<string> rawTags = new List<string>();
        int pos = 0;
        while (true)
        {
            int next = subString.IndexOf("||", pos, StringComparison.Ordinal);
            if (next < 0)
            {
                rawTags.Add(subString.Substring(pos));
                break;
            }

            rawTags.Add(subString.Substring(pos, next - pos));
            pos = next + 2;
        }

        while (rawTags.Count > 0 && rawTags[rawTags.Count - 1].Length == 0)
        {
            rawTags.RemoveAt(rawTags.Count - 1);
        }

        if (rawTags.Count == 0)
        {
            // Java: throw new Exception("subString split error")
            throw new ArgumentException("subString split error");
        }

        foreach (string raw in rawTags)
        {
            string tag = TrimTag(raw);
            if (tag.Length > 0)
            {
                sub.TagsSet.Add(tag);
                sub.CodeSet.Add(JavaHash.JavaStringHash(tag));
            }
        }

        return sub;
    }

    private static string TrimTag(string s)
    {
        int b = 0;
        while (b < s.Length && IsTagWs(s[b]))
        {
            b++;
        }

        int e = s.Length - 1;
        while (e >= b && IsTagWs(s[e]))
        {
            e--;
        }

        return b > e ? string.Empty : s.Substring(b, e - b + 1);
    }

    private static bool IsTagWs(char c) => c == ' ' || c == '\t' || c == '\r' || c == '\n';
}

File: cpp/src/common/subscription_data.cpp
// SubscriptionData / FilterAPI 的实现。
#include "rocketmq/common/subscription_data.h"

#include <cstdint>
#include <string>

#include "rocketmq/common/byte_buffer.h"
#include "rocketmq/common/util_all.h"

namespace rocketmq {

const char* const ExpressionType::TAG = "TAG";
const char* const ExpressionType::SQL92 = "SQL92";
const char* const ExpressionType::CLASS_FILTER = "CLASS_FILTER";

namespace {

// Java Set.hashCode() = 各元素 hashCode 之和（32 位有符号回绕）
int32_t setHashCode(const std::set<std::string>& s) {
    int64_t h = 0;
    for (const std::string& e : s) {
        h += javaStringHash(e);
        h &= 0xFFFFFFFFL;
    }
    return static_cast<int32_t>(h);
}

int32_t setHashCode(const std::set<int64_t>& s) {
    int64_t h = 0;
    for (int64_t e : s) {
        h += e;
        h &= 0xFFFFFFFFL;
    }
    return static_cast<int32_t>(h);
}

// Java set 序列化为 JSON 数组
JsonValue setToJson(const std::set<std::string>& s) {
    JsonValue arr = JsonValue::makeArray();
    for (const std::string& e : s) {
        arr.pushArray(JsonValue::makeString(e));
    }
    return arr;
}

JsonValue setToJson(const std::set<int64_t>& s) {
    JsonValue arr = JsonValue::makeArray();
    for (int64_t e : s) {
        arr.pushArray(JsonValue::makeInt(e));
    }
    return arr;
}

}  // namespace

SubscriptionData::SubscriptionData() : subVersion(UtilAll::currentTimeMillis()) {}

SubscriptionData::SubscriptionData(const std::string& topic, const std::string& subString)
    : topic(topic), subString(subString), subVersion(UtilAll::currentTimeMillis()) {}

JsonValue SubscriptionData::toJson() const {
    // 键按字母序，与 fastjson2 输出一致
    JsonValue o = JsonValue::makeObject();
    o.set("classFilterMode", JsonValue::makeBool(classFilterMode));
    o.set("codeSet", setToJson(codeSet));
    o.set("expressionType", JsonValue::makeString(expressionType));
    o.set("subString", JsonValue::makeString(subString));
    o.set("subVersion", JsonValue::makeInt(subVersion));
    o.set("tagsSet", setToJson(tagsSet));
    o.set("topic", JsonValue::makeString(topic));
    return o;
}

SubscriptionData SubscriptionData::fromJson(const JsonValue& v) {
    SubscriptionData sd;
    if (!v.isObject()) {
        return sd;
    }
    bool b = false;
    if (v.tryGetBool("classFilterMode", b)) {
        sd.classFilterMode = b;
    }
    std::string s;
    if (v.tryGetString("expressionType", s)) {
        sd.expressionType = s;
    }
    if (v.tryGetString("subString", s)) {
        sd.subString = s;
    }
    if (v.tryGetString("topic", s)) {
        sd.topic = s;
    }
    int64_t n = 0;
    if (v.tryGetInt("subVersion", n)) {
        sd.subVersion = n;
    }
    const JsonValue* tags = v.find("tagsSet");
    if (tags != nullptr && tags->isArray()) {
        for (size_t i = 0; i < tags->size(); ++i) {
            sd.tagsSet.insert(tags->at(i).stringValue());
        }
    }
    const JsonValue* codes = v.find("codeSet");
    if (codes != nullptr && codes->isArray()) {
        for (size_t i = 0; i < codes->size(); ++i) {
            sd.codeSet.insert(codes->at(i).intValue());
        }
    }
    return sd;
}

bool SubscriptionData::operator==(const SubscriptionData& o) const {
    return classFilterMode == o.classFilterMode && codeSet == o.codeSet
        && subString == o.subString && subVersion == o.subVersion && tagsSet == o.tagsSet
        && topic == o.topic && expressionType == o.expressionType;
}

int SubscriptionData::compareTo(const SubscriptionData& o) const {
    const std::string a = topic + "@" + subString;
    const std::string b = o.topic + "@" + o.subString;
    if (a == b) {
        return 0;
    }
    return a < b ? -1 : 1;
}

int SubscriptionData::hashCode() const {
    int64_t result = 1;
    result = 31 * result + (classFilterMode ? 1231 : 1237);
    result = 31 * result + setHashCode(codeSet);
    result = 31 * result + javaStringHash(subString);
    result = 31 * result + setHashCode(tagsSet);
    result = 31 * result + javaStringHash(topic);
    result = 31 * result + javaStringHash(expressionType);
    result &= 0xFFFFFFFFL;
    return static_cast<int32_t>(result);
}

std::string SubscriptionData::toString() const {
    return "SubscriptionData [classFilterMode=" + std::string(classFilterMode ? "true" : "false")
         + ", topic=" + topic + ", subString=" + subString + ", tagsSet=" + setToJson(tagsSet).dump()
         + ", codeSet=" + setToJson(codeSet).dump() + ", subVersion=" + std::to_string(subVersion)
         + ", expressionType=" + expressionType + "]";
}

SubscriptionData FilterAPI::buildSubscriptionData(const std::string& topic,
                                                  const std::string& subString) {
    SubscriptionData sub(topic, subString);
    // Java `FilterAPI.buildSubscriptionData` 逐句对齐（探针 /tmp/subprobe/SubProbe.java 与
    // BlankProbe.java 实测过四种输入）：
    //
    //   StringUtils.isEmpty(subString) || subString.equals("*")  → setSubString("*") 后**直接 return**
    //       ⇒ tagsSet 与 codeSet 都保持空
    //   否则按 "||" 做 Java String.split（**丢弃末尾空串**），然后 trim、丢空片段，
    //       每个 tag 同时进 tagsSet 与 codeSet（codeSet 存 tag.hashCode()）。
    //       Java-split 结果数组长度为 0 时（即 subString 形如 "||"、"||||"）抛 Exception。
    //
    // ⚠ 两处曾有的偏差（都会污染心跳、并让客户端二次 tag 过滤失效）：
    //   ① 给 "*" 塞 tagsSet={"*"} —— Java 里 tagsSet 非空才是"客户端二次 tag 过滤"的开关
    //      （PullAPIWrapper.processPullResult 的 `!tagsSet.isEmpty()`），塞了 "*" 会让订阅
    //      全量时把所有正常 tag 的消息客户端自己过滤掉；
    //   ② 从不填 codeSet —— 它是 broker 侧按 tag 哈希过滤的依据
    //      （ExpressionMessageFilter.isMatchedByConsumeQueue 走 codeSet.contains）。
    // 另注：判空必须用"空串"而不是"全空白"—— Java StringUtils.isEmpty 只认 null/""，
    // 纯空白（如 "   "）会走进 split 分支，结果 tagsSet 空但 subString 原样保留。
    if (subString.empty() || subString == "*") {
        sub.subString = "*";
        return sub;
    }
    // Java String.split("\\|\\|")：先全切，再丢弃**末尾**空串
    std::vector<std::string> rawTags;
    size_t pos = 0;
    while (true) {
        size_t next = subString.find("||", pos);
        if (next == std::string::npos) {
            rawTags.push_back(subString.substr(pos));
            break;
        }
        rawTags.push_back(subString.substr(pos, next - pos));
        pos = next + 2;
    }
    while (!rawTags.empty() && rawTags.back().empty()) {
        rawTags.pop_back();
    }
    if (rawTags.empty()) {
        // Java: throw new Exception("subString split error")
        throw std::invalid_argument("subString split error");
    }
    for (std::string& tag : rawTags) {
        size_t b = tag.find_first_not_of(" \t\r\n");
        size_t e = tag.find_last_not_of(" \t\r\n");
        tag = (b == std::string::npos) ? std::string() : tag.substr(b, e - b + 1);
        if (!tag.empty()) {
            sub.tagsSet.insert(tag);
            sub.codeSet.insert(static_cast<int64_t>(javaStringHash(tag)));
        }
    }
    return sub;
}

}  // namespace rocketmq
